RFC process (#174)
cloudfoundry/community: toc/rfc/README.md

# Cloud Foundry Request For Comments

Cloud Foundry community members use this directory as a forum to make, discuss and record both process and technical decisions. The outcomes of these discussions can be either an Action Plan, a new Standard or Process that the Cloud Foundry community should follow. This repository is open as a reference for all working groups and for the wider community.

## Process

1. Create a new branch on this repo and copy `rfc-000-template.md` to `rfc-000-my-proposal.md` and edit.

> 1. If your RFC is only relevant for a single Working Group, please place it in a directory for the working group.

Reviewer: Is this directory supposed to be the designated directory for the working group under `toc/working-groups`, or an identically named directory under `toc/rfc`, or something else?

Author: I've added some clarifying text

>    1. If your RFC is only relevant for a sigle Working Group, and the Working Group hosts RFCs elsewhere, please create a directory and place a link to where the RFCs can be found to aid in discoverability and follow the already agreed process. 

Reviewer 2: Most of the time, the directory will already exist and should already point to the working group's RFC process. Maybe, we'll just have working groups file an RFC here that states the working group has a separate RFC process and provides a link to it. WDYT?
```suggestion
```

Author: How about 

```
If your RFC is only relevant for a single Working Group, and the Working Group hosts RFCs elsewhere and this step has not already been completed, please create a directory and place a link to where the RFCs can be found to aid in discoverability and follow the already agreed process.
```

>    1. If your RFC is only relevant for a sigle Working Group, and the Working Group hosts RFCs elsewhere and this step has not already been completed, please create a directory within toc/rfc and place a link to where the RFCs can be found to aid in discoverability and follow the already agreed process. 

Reviewer: ```suggestion
   1. If that Working Group hosts its RFCs elsewhere, its RFC subdirectory will contain a file called `EXTERNAL.md` that contains a link to that group's RFC repository. In that case, follow that link to create the group-specific RFC in that repository.
```
This external-redirect file seems like something that the TOC can backfill for existing WGs (which I expect is only Paketo) and can create contemporaneously with any new or altered WGs.

3. Include any images etc in a separate directory named `rfc-000` and link to them.

> 4. Make a Pull Request (PR) for your branch.
> 5. Rename your file and directory with the number of the PR and push as a new commits.

Reviewer 2: ```suggestion
1. Make a Pull Request (PR) for your branch.
1. Add a link to the readable version of the RFC to the PR decription.
1. Rename your file and directory with the number of the PR and push as a new commits.
```

> 1. Rename your file and directory with the number of the PR and push as a new commits.

Author: It should be the next sequential number. I'll edit the text to make that clearer

1. Tag `@toc` and any relevant working groups in your PR to enable discussion of the RFC. 
1. Cloud Foundry community members discuss your proposal using both inline comments against your RFC document and the general PR comments section.
1. As changes are requested and agreed in comments, make the changes in your RFC and push them as new commits.

> 1. Stay active in the discussion and encourage and remind other relevant people to participate. If you’re unsure who should be involved in a discussion, ask your Technical Lead, Execution Lead or the TOC. If you start an RFC it’s up to you to push it through the process and engage people.

Reviewer: ```suggestion
1. Stay active in the discussion and encourage and remind other relevant people to participate. If you’re unsure who should be involved in a discussion, ask the Leads for relevant working groups or the TOC. If you start an RFC it is up to you to engage people to guide it through the process.
```

> 1. Once consensus is reached and approvals given using the Github review system, the PR can be merged.

Reviewer: This stage might also benefit from a final comment period like the one in the CNB process (https://github.com/buildpacks/rfcs#final-comment-period). That final comment period and the motion to enter it could also be a common process for acceptance and rejection, since both require consensus of the relevant governance body (TOC or WG).

1. When an RFC is accepted, ensure the Cloud Foundry community is made aware of it via Slack using the previously used channels.

> 1. An RFC can be rejected. This can happen if a consensus isn’t reached, or people agree rejecting it is the right thing to do. In this case the PR should be closed with a suitable comment.

Reviewer: For the first case, should there be a "timeout" duration for which lack of consensus means that the RFC is automatically rejected? Or should we say that the RFC author (or perhaps the TOC) should explicitly acknowledge that there is a failure to reach consensus and then to close the RFC PR?

Also, who has the authority to close a rejected RFC PR? The proposal author and the TOC (assuming its own consensus)?

Author: Hopefully this should now be resolved

> 1. Create a new branch on this repo and copy `rfc-000-template.md` to `rfc-000-my-proposal.md` and edit.
> 2. If your RFC is only relevant for a single Working Group, please place it in a directory for the working group.
>    1. If your RFC is only relevant for a sigle Working Group, and the Working Group hosts RFCs elsewhere, please create a directory and place a link to where the RFCs can be found to aid in discoverability and follow the already agreed process. 
> 3. Include any images etc in a separate directory named `rfc-000` and link to them.
> 4. Make a Pull Request (PR) for your branch.
> 5. Rename your file and directory with the number of the PR and push as a new commits.
> 6. Tag `@toc` and any relevant working groups in your PR to enable discussion of the RFC. 
> 7. Cloud Foundry community members discuss your proposal using both inline comments against your RFC document and the general PR comments section.
> 8. As changes are requested and agreed in comments, make the changes in your RFC and push them as new commits.
> 9. Stay active in the discussion and encourage and remind other relevant people to participate. If you’re unsure who should be involved in a discussion, ask your Technical Lead, Execution Lead or the TOC. If you start an RFC it’s up to you to push it through the process and engage people.
> 10. Once consensus is reached and approvals given using the Github review system, the PR can be merged.
> 11. When an RFC is accepted, ensure the Cloud Foundry community is made aware of it via Slack using the previously used channels.
> 12. An RFC can be rejected. This can happen if a consensus isn’t reached, or people agree rejecting it is the right thing to do. In this case the PR should be closed with a suitable comment.

Reviewer 2: ```suggestion
1. Create a new branch on this repo and copy `rfc-000-template.md` to `rfc-000-my-proposal.md` and edit.
1. If your RFC is only relevant for a single Working Group, please place it in a directory for the working group.
   1. If your RFC is only relevant for a sigle Working Group, and the Working Group hosts RFCs elsewhere, please create a directory and place a link to where the RFCs can be found to aid in discoverability and follow the already agreed process. 
1. Include any images etc in a separate directory named `rfc-000` and link to them.
1. Make a Pull Request (PR) for your branch.
1. Rename your file and directory with the number of the PR and push as a new commits.
1. Tag `@toc` and any relevant working groups in your PR to enable discussion of the RFC. 
1. Cloud Foundry community members discuss your proposal using both inline comments against your RFC document and the general PR comments section.
1. As changes are requested and agreed in comments, make the changes in your RFC and push them as new commits.
1. Stay active in the discussion and encourage and remind other relevant people to participate. If you’re unsure who should be involved in a discussion, ask your Technical Lead, Execution Lead or the TOC. If you start an RFC it’s up to you to push it through the process and engage people.
1. Once consensus is reached and approvals given using the Github review system, the PR can be merged.
1. When an RFC is accepted, ensure the Cloud Foundry community is made aware of it via Slack using the previously used channels.
1. An RFC can be rejected. This can happen if a consensus isn’t reached, or people agree rejecting it is the right thing to do. In this case the PR should be closed with a suitable comment.
```

# Cloud Foundry Request For Comments

> Cloud Foundry community members use this directory as a forum to discuss, make and record both process and technical decisions. The outcomes of these discussions can be either an Action Plan, a new Standard or Process that the Cloud Foundry community should follow. This repository is open as a reference for all working groups and for the wider community.

Reviewer 2: ```suggestion
Cloud Foundry community members use this directory as a forum to make, discuss and record both process and technical decisions. The outcomes of these discussions can be either an Action Plan, a new Standard or Process that the Cloud Foundry community should follow. This repository is open as a reference for all working groups and for the wider community.
```

1. Create a new RFC PR as above, noting in the summary which RFC it is superseding.
2. In the same branch, mark the old RFC as superseded and link to the new RFC and move (using `git mv`) it into the archived directory.
3. When the new RFC is accepted and the PR is merged, the old RFC will no longer be active.

## Managing Action Plans
For RFCs where the outcome is an agreed Action Plan, you may want to update the RFC with meaningful status updates in new PRs. Once the plan is either complete or no longer relevant, it should be moved to the archived directory in a new PR.

## Managing Processes

Process RFCs shouldn’t be substantially altered after they are accepted, although it’s fine to correct typos and other mistakes via a new PR. In order to change a Process, the original RFC must be superseded by a new one. The process for this is:

1. Create a new RFC PR as above, noting in the summary which RFC it is superseding.
2. In the same branch, mark the old RFC as superseded and link to the new RFC and move (using `git mv`) it into the archived directory.

> 3. When the new RFC is accepted and the PR is merged, the old RFC will no longer be active.

Reviewer 2: This looks pretty duplicative to "Managing Standards". Should we combine this to e.g. "Managing RFCs"?
